Civi/Eck/API/Entity.php

```diff
@@ -96,6 +101,70 @@ public static function afformEntityTypes(GenericHookEvent $e) {
     }
   }
 
+  /**
+   * Generates afforms for each ECK entity type and sub-type.
+   *
+   * @param \Civi\Core\Event\GenericHookEvent $event
+   * @throws \API_Exception
+   */
+  public static function getEckAfforms($event) {
+    // Early return if forms are not requested
+    if ($event->getTypes && !in_array('form', $event->getTypes, TRUE)) {
+      return;
+    }
+
+    $afforms =& $event->afforms;
+    $getNames = $event->getNames;
+
+    // Early return if this api call is fetching afforms by name and those names are not eck-related
+    if (
+      (!empty($getNames['name']) && !strstr(implode(' ', $getNames['name']), 'afformEck_'))
+      || (!empty($getNames['module_name']) && !strstr(implode(' ', $getNames['module_name']), 'afformEck'))
+      || (!empty($getNames['directive_name']) && !strstr(implode(' ', $getNames['directive_name']), 'afform-eck'))
+    ) {
+      return;
+    }
+
+    foreach (\CRM_Eck_BAO_EckEntityType::getEntityTypes() as $entityType) {
+      foreach (\CRM_Eck_BAO_EckEntityType::getSubTypes($entityType['name'], FALSE) as $subType) {
+        $name = 'afform' . $entityType['entity_name'] . '_' . $subType['name'];
+        $item = [
+          'name' => $name,
+          'type' => 'form',
+          'title' => $entityType['label'] . ' (' . $subType['label'] . ')',
+          'description' => '',
+          'is_dashlet' => FALSE,
+          'is_public' => FALSE,
+          'is_token' => FALSE,
+          'permission' => 'access CiviCRM',
+          'server_route' => "civicrm/eck/edit/{$entityType['name']}/{$subType['name']}",
```

Comment:
Afform routes don't seem to allow spaces and special characters, which might be included in subtype names. I'd suggest making subtype names match those restrictions so that this part of the code can stay the same. Maybe a test should be added with a subtype having a space in its label.